// typescript-lit-html-plugin/lib/html-template-language-service.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
//
// Original code forked from https://github.com/Quramy/ts-graphql-plugin
Object.defineProperty(exports, "__esModule", { value: true });
var vscode_html_languageservice_1 = require("vscode-html-languageservice");
var vscode = require("vscode-languageserver-types");
function arePositionsEqual(left, right) {
    return left.line === right.line && left.character === right.character;
}
var CompletionsCache = /** @class */ (function () {
    function CompletionsCache() {
    }
    CompletionsCache.prototype.getCached = function (context, position) {
        if (this._completions
            && context.fileName === this._cachedCompletionsFile
            && this._cachedCompletionsPosition && arePositionsEqual(position, this._cachedCompletionsPosition)
            && context.text === this._cachedCompletionsContent) {
            return this._completions;
        }
        return undefined;
    };
    CompletionsCache.prototype.updateCached = function (context, position, completions) {
        this._cachedCompletionsFile = context.fileName;
        this._cachedCompletionsPosition = position;
        this._cachedCompletionsContent = context.text;
        this._completions = completions;
    };
    return CompletionsCache;
}());
var HtmlTemplateLanguageService = /** @class */ (function () {
    function HtmlTemplateLanguageService(typescript, configuration, logger) {
        this.typescript = typescript;
        this.configuration = configuration;
        this.logger = logger;
        this._completionsCache = new CompletionsCache();
    }
    Object.defineProperty(HtmlTemplateLanguageService.prototype, "htmlLanguageService", {
        get: function () {
            if (!this._htmlLanguageService) {
                this._htmlLanguageService = vscode_html_languageservice_1.getLanguageService();
            }
            return this._htmlLanguageService;
        },
        enumerable: true,
        configurable: true
    });
    HtmlTemplateLanguageService.prototype.getCompletionsAtPosition = function (context, position) {
        var items = this.getCompletionItems(context, position);
        return translateCompletionItemsToCompletionInfo(this.typescript, items);
    };
    HtmlTemplateLanguageService.prototype.getCompletionEntryDetails = function (context, position, name) {
        var items = this.getCompletionItems(context, position).items;
        var item = items.find(function (x) { return x.label === name; });
        if (!item) {
            return {
                name: name,
                kind: this.typescript.ScriptElementKind.unknown,
                kindModifiers: '',
                tags: [],
                displayParts: toDisplayParts(name),
                documentation: [],
            };
        }
        return translateCompletionItemsToCompletionEntryDetails(this.typescript, item);
    };
    HtmlTemplateLanguageService.prototype.getQuickInfoAtPosition = function (context, position) {
        var doc = this.createVirtualDocument(context);
        var htmlDoc = this.htmlLanguageService.parseHTMLDocument(doc);
        var hover = this.htmlLanguageService.doHover(doc, position, htmlDoc);
        if (hover) {
            return this.translateHover(hover, position, context);
        }
        return undefined;
    };
    HtmlTemplateLanguageService.prototype.getFormattingEditsForRange = function (context, start, end, settings) {
        var _this = this;
        if (!this.configuration.format.enabled) {
            return [];
        }
        var doc = this.createVirtualDocument(context);
        var htmlDoc = this.htmlLanguageService.parseHTMLDocument(doc);
        // Make sure we don't get rid of leading newline
        var leading = context.text.match(/^\s*\n/);
        if (leading) {
            start += leading[0].length;
        }
        // or any trailing newlines
        var trailing = context.text.match(/\n\s*$/);
        if (trailing) {
            end -= trailing[0].length;
        }
        if (end <= start) {
            return [];
        }
        var range = this.toVsRange(context, start, end);
        var edits = this.htmlLanguageService.format(doc, range, {
            tabSize: settings.tabSize,
            insertSpaces: !!settings.convertTabsToSpaces,
            wrapLineLength: 120,
            unformatted: '',
            contentUnformatted: 'pre,code,textarea',
            indentInnerHtml: false,
            preserveNewLines: true,
            maxPreserveNewLines: null,
            indentHandlebars: false,
            endWithNewline: false,
            extraLiners: 'head, body, /html',
            wrapAttributes: 'auto',
        });
        return edits.map(function (vsedit) {
            return {
                span: _this.toTsSpan(context, vsedit.range),
                newText: vsedit.newText,
            };
        });
    };
    HtmlTemplateLanguageService.prototype.toVsRange = function (context, start, end) {
        return {
            start: context.toPosition(start),
            end: context.toPosition(end),
        };
    };
    HtmlTemplateLanguageService.prototype.toTsSpan = function (context, range) {
        var editStart = context.toOffset(range.start);
        var editEnd = context.toOffset(range.end);
        return {
            start: editStart,
            length: editEnd - editStart,
        };
    };
    HtmlTemplateLanguageService.prototype.createVirtualDocument = function (context) {
        var contents = context.text;
        return {
            uri: 'untitled://embedded.html',
            languageId: 'html',
            version: 1,
            getText: function () { return contents; },
            positionAt: function (offset) {
                return context.toPosition(offset);
            },
            offsetAt: function (p) {
                return context.toOffset(p);
            },
            lineCount: contents.split(/n/g).length + 1,
        };
    };
    HtmlTemplateLanguageService.prototype.getCompletionItems = function (context, position) {
        var cached = this._completionsCache.getCached(context, position);
        if (cached) {
            return cached;
        }
        var doc = this.createVirtualDocument(context);
        var htmlDoc = this.htmlLanguageService.parseHTMLDocument(doc);
        var completions = this.htmlLanguageService.doComplete(doc, position, htmlDoc);
        this._completionsCache.updateCached(context, position, completions);
        return this.htmlLanguageService.doComplete(doc, position, htmlDoc);
    };
    HtmlTemplateLanguageService.prototype.translateHover = function (hover, position, context) {
        var contents = [];
        var convertPart = function (hoverContents) {
            if (typeof hoverContents === 'string') {
                contents.push({ kind: 'unknown', text: hoverContents });
            }
            else if (Array.isArray(hoverContents)) {
                hoverContents.forEach(convertPart);
            }
            else {
                contents.push({ kind: 'unknown', text: hoverContents.value });
            }
        };
        convertPart(hover.contents);
        var start = context.toOffset(hover.range ? hover.range.start : position);
        return {
            kind: this.typescript.ScriptElementKind.unknown,
            kindModifiers: '',
            textSpan: {
                start: start,
                length: hover.range ? context.toOffset(hover.range.end) - start : 1,
            },
            displayParts: [],
            documentation: contents,
            tags: [],
        };
    };
    return HtmlTemplateLanguageService;
}());
exports.default = HtmlTemplateLanguageService;
function translateCompletionItemsToCompletionInfo(typescript, items) {
    return {
        isGlobalCompletion: false,
        isMemberCompletion: false,
        isNewIdentifierLocation: false,
        entries: items.items.map(function (x) { return translateCompetionEntry(typescript, x); }),
    };
}
function translateCompletionItemsToCompletionEntryDetails(typescript, item) {
    return {
        name: item.label,
        kindModifiers: 'declare',
        kind: item.kind ? translateionCompletionItemKind(typescript, item.kind) : typescript.ScriptElementKind.unknown,
        displayParts: toDisplayParts(item.detail),
        documentation: toDisplayParts(item.documentation),
        tags: [],
    };
}
function translateCompetionEntry(typescript, item) {
    return {
        name: item.label,
        kindModifiers: '',
        kind: item.kind ? translateionCompletionItemKind(typescript, item.kind) : typescript.ScriptElementKind.unknown,
        sortText: '0',
    };
}
function translateionCompletionItemKind(typescript, kind) {
    switch (kind) {
        case vscode.CompletionItemKind.Method:
            return typescript.ScriptElementKind.memberFunctionElement;
        case vscode.CompletionItemKind.Function:
            return typescript.ScriptElementKind.functionElement;
        case vscode.CompletionItemKind.Constructor:
            return typescript.ScriptElementKind.constructorImplementationElement;
        case vscode.CompletionItemKind.Field:
        case vscode.CompletionItemKind.Variable:
            return typescript.ScriptElementKind.variableElement;
        case vscode.CompletionItemKind.Class:
            return typescript.ScriptElementKind.classElement;
        case vscode.CompletionItemKind.Interface:
            return typescript.ScriptElementKind.interfaceElement;
        case vscode.CompletionItemKind.Module:
            return typescript.ScriptElementKind.moduleElement;
        case vscode.CompletionItemKind.Property:
            return typescript.ScriptElementKind.memberVariableElement;
        case vscode.CompletionItemKind.Unit:
        case vscode.CompletionItemKind.Value:
            return typescript.ScriptElementKind.constElement;
        case vscode.CompletionItemKind.Enum:
            return typescript.ScriptElementKind.enumElement;
        case vscode.CompletionItemKind.Keyword:
            return typescript.ScriptElementKind.keyword;
        case vscode.CompletionItemKind.Color:
            return typescript.ScriptElementKind.constElement;
        case vscode.CompletionItemKind.Reference:
            return typescript.ScriptElementKind.alias;
        case vscode.CompletionItemKind.File:
            return typescript.ScriptElementKind.moduleElement;
        case vscode.CompletionItemKind.Snippet:
        case vscode.CompletionItemKind.Text:
        default:
            return typescript.ScriptElementKind.unknown;
    }
}
function toDisplayParts(text) {
    return text ? [{ text: text, kind: 'text' }] : [];
}
